### PR TITLE
Get joypad's vendor ID, product ID and name on Windows for XInput devices.

### PR DESCRIPTION
**Pull request by MJacred**
Currently, XInput devices (Windows only) lack info on vendor, product and its name. DirectInput also does not provide vendor and product. This PR is to fix that.

TODO
* [x] vendor ID (both)
* [x] product ID (both)
* [x] name (XInput only)

NOTES
* Previously, all XInput joypads had the name `XInput Gamepad`, now they have proper names
* I'm stripping "Controller (…)" from the name. Also for DirectInput devices. There's no reason for keeping this redundancy.

Ideas for alternative implementation
* using `joyGetDevCapsW`, therefore loading Winmm.dll
  * I haven't tested it, but it should also be possible to get product and vendor ID using `GetRawInputDeviceInfoA` or `DIDEVICEINSTANCE`, basically making loading Winmm.dll obsolete. If I should check it out, I will. I'm not doing that right now, because I'm lacking background logic of the whole device mess (these APIs are all deprecated anyway...)
* I also could have generated a guid, as the DirectInput code does, but I cannot verify if they are correct…
  * `sprintf_s(uid, "%04x%04x%04x%04x%04x%04x%04x%04x", BSWAP16(0x03), 0, jc.wMid, 0, jc.wPid, 0, 0, 0);` generated `03000000045e0000028e000000000000` and `03000000046d0000c21f000000000000`

Testproject for this PR:  
[xinput_info.zip](https://github.com/user-attachments/files/18142511/xinput_info.zip)

output of Testproject on Windows
finally

### CODE REVIEW

**Comment by MJacred** [comment]
OK, so the Godot build artifact for Windows crashes on startup. As it's not a debug build, there's no stacktrace. And I can't get Godot to build on Windows, because it just can't find mingw (yes, I followed Godot's build instructions).  
I'll try finding help on Discord over the weekend. If nothing helps, I'll try cross-compilation from Linux...

**Comment by Calinou** [comment]
@MJacred 64-bit Windows editor binary with MinGW debug symbols (in a separate file): https://fromsmash.com/Godot-PR-test-GH-98861

**Comment by MJacred** [comment]
@Calinou: I think I found something: "3221226525", which is apparently an access violation..

As Godot only crashes when I plugin an XInput device, I suspect the issue lies in `JoypadWindows::probe_joypads`. I noticed I did a `sizeof(jc)` instead of `sizeof(JOYCAPS)` (fixed that just now), but I don't think that was it. It might be `String(jc.szPname)`, where `szPname` is a `char[32]`...

---

This is the command I used
```sh
gdb --symbols=godot.windows.editor.x86_64.exe.debugsymbols --exec=godot.windows.editor.x86_64.console.exe
```
but I got
```
Reading symbols from C:...\godot.windows.editor.x86_64.exe.debugsymbols...
C:...\godot.windows.editor.x86_64.exe.debugsymbols: Invalid argument.
```

I also tried without `symbols`, but then gdb read symbols from normal exe…  
So, I'm not sure what the best way is regarding debug symbols...

Regarding the 2 different cases, I could only find unresolved issues for error code "3221226525". But I found sth. regarding "3221225477".


## Starting Godot with XInput device plugged in

**found info on error code**

```sh
Starting program: C:...\godot.windows.editor.x86_64.console.exe
[New Thread 14628.0x315c]
[New Thread 14628.0x170c]
[New Thread 14628.0x3bc4]
Godot Engine v4.4.dev.custom_build.949f572c7 (2024-11-05 20:27:39 UTC) - https://godotengine.org

================================================================
CrashHandlerException: Program crashed with signal 11
Engine version: Godot Engine v4.4.dev.custom_build (949f572c70cfc15b0711ba23bd745874d8038d2f)
Dumping the backtrace. Please include this when reporting the bug to the project developer.
[1] error(-1): no debug info in PE/COFF executable
[2] error(-1): no debug info in PE/COFF executable
[3] error(-1): no debug info in PE/COFF executable
[4] error(-1): no debug info in PE/COFF executable
[5] error(-1): no debug info in PE/COFF executable
[6] error(-1): no debug info in PE/COFF executable
-- END OF BACKTRACE --
================================================================
[Thread 14628.0x170c exited with code 3221225477]
[Thread 14628.0x3bc4 exited with code 3221225477]
[Thread 14628.0x315c exited with code 3221225477]

Program terminated with signal SIGSEGV, Segmentation fault.
The program no longer exists.
```


## Starting Godot, then plugging in XInput device

**meaning of error code unknown**

```sh
Starting program: C:...\godot.windows.editor.x86_64.console.exe
[New Thread 6944.0x2ba0]
[New Thread 6944.0x2a88]
[New Thread 6944.0x920]
Godot Engine v4.4.dev.custom_build.949f572c7 (2024-11-05 20:27:39 UTC) - https://godotengine.org
OpenGL API 3.3.0 NVIDIA 560.94 - Compatibility - Using Device: NVIDIA - NVIDIA GeForce GTX 970

================================================================
CrashHandlerException: Program crashed with signal 11
Engine version: Godot Engine v4.4.dev.custom_build (949f572c70cfc15b0711ba23bd745874d8038d2f)
Dumping the backtrace. Please include this when reporting the bug to the project developer.
[1] error(-1): no debug info in PE/COFF executable
[2] error(-1): no debug info in PE/COFF executable
[3] error(-1): no debug info in PE/COFF executable
[4] error(-1): no debug info in PE/COFF executable
[5] error(-1): no debug info in PE/COFF executable
[6] error(-1): no debug info in PE/COFF executable
[7] error(-1): no debug info in PE/COFF executable
[8] error(-1): no debug info in PE/COFF executable
-- END OF BACKTRACE --
================================================================
[Thread 6944.0x920 exited with code 3221226525]
[Thread 6944.0x2a88 exited with code 3221226525]
[Thread 6944.0x2ba0 exited with code 3221226525]
[Inferior 1 (process 6944) exited with code 030000002035]
```

**Comment by MJacred** [comment]
Current status: it would work if I used `joyGetDevCaps` directly, like `midiInGetDevCaps` from `drivers\winmidi\midi_driver_winmidi.cpp`. So... I'm doing sth. wrong regarding `joyGetDevCaps_t`. No idea what excatly...

**So help here is more than appreciated.**

So, the result I got when using the func directly was this:
get_xinfo

As you can see, the device name is disappointing. I know that Windows knows the name. While the device manager has no clue, the game controllers settings do.

windows_devices

Right now, I see 2 options on how to get the name:
1. Make use of `JoypadWindows::enumCallback` -> with winmm, we get product and vendor ID; then fetch them again using DirectInput, wich also gets us the name (additionally to the same IDs). Then simply map back.
2. Reading joypads as devices and then using some generic Windows api for reading device name (should work, but right now no idea how)

**Comment by MJacred** [comment]
Switched to fetching the func `joyGetDevCapsW` instead of the macro `joyGetDevCaps`, because it seems it's impossible with the macro. As *W and *A funcs are used in the engine all over the place, I'll stick to *W here.

I'll tackle getting the name next (but not this weekend).

**Comment by MJacred** [comment]
OK, I got the name - ready for review

File: doc/classes/Input.xml [hunk 1]
```diff
@@ -118,7 +118,7 @@
 			<return type="String" />
 			<param index="0" name="device" type="int" />
 			<description>
-				Returns an SDL2-compatible device GUID on platforms that use gamepad remapping, e.g. [code]030000004c050000c405000000010000[/code]. Returns [code]"Default Gamepad"[/code] otherwise. Godot uses the [url=https://github.com/gabomdq/SDL_GameControllerDB]SDL2 game controller database[/url] to determine gamepad names and mappings based on this GUID.
+				Returns an SDL2-compatible device GUID on platforms that use gamepad remapping, e.g. [code]030000004c050000c405000000010000[/code]. Returns an empty String otherwise. Godot uses the [url=https://github.com/gabomdq/SDL_GameControllerDB]SDL2 game controller database[/url] to determine gamepad names and mappings based on this GUID. Although Godot overwrites [url=https://github.com/godotengine/godot/blob/master/core/input/godotcontrollerdb.txt]some of the GUIDs[/url]. DirectX joypads on Windows, for example, all return the GUID [code]__XINPUT_DEVICE__[/code].
```
Comment (Calinou) [inline doc/classes/Input.xml:121]:
```suggestion
				Returns an SDL2-compatible device GUID on platforms that use gamepad remapping, e.g. [code]030000004c050000c405000000010000[/code]. Returns an empty String otherwise. Godot uses the [url=https://github.com/gabomdq/SDL_GameControllerDB]SDL2 game controller database[/url] to determine gamepad names and mappings based on this GUID, although Godot overwrites [url=https://github.com/godotengine/godot/blob/master/core/input/godotcontrollerdb.txt]some of the GUIDs[/url]. DirectX joypads on Windows, for example, all return the GUID [code]__XINPUT_DEVICE__[/code].
```

File: doc/classes/Input.xml [hunk 2]
```diff
@@ -140,7 +143,8 @@
 			<return type="String" />
 			<param index="0" name="device" type="int" />
 			<description>
-				Returns the name of the joypad at the specified device index, e.g. [code]PS4 Controller[/code]. Godot uses the [url=https://github.com/gabomdq/SDL_GameControllerDB]SDL2 game controller database[/url] to determine gamepad names.
+				Returns the name of the joypad at the specified device index, e.g. [code]PS4 Controller[/code]. Godot uses the [url=https://github.com/gabomdq/SDL_GameControllerDB]SDL2 game controller database[/url] to determine gamepad names. Although Godot overwrites [url=https://github.com/godotengine/godot/blob/master/core/input/godotcontrollerdb.txt]some of the names[/url], especially on Windows for XInput devices.
```
Comment (Calinou) [inline doc/classes/Input.xml:146]:
```suggestion
				Returns the name of the joypad at the specified device index, e.g. [code]PS4 Controller[/code]. Godot uses the [url=https://github.com/gabomdq/SDL_GameControllerDB]SDL2 game controller database[/url] to determine gamepad names, although Godot overwrites [url=https://github.com/godotengine/godot/blob/master/core/input/godotcontrollerdb.txt]some of the names[/url], especially on Windows for XInput devices.
```

File: doc/classes/Input.xml [hunk 3]
```diff
@@ -127,7 +127,10 @@
 			<description>
 				Returns a dictionary with extra platform-specific information about the device, e.g. the raw gamepad name from the OS or the Steam Input index.
 				On Windows the dictionary contains the following fields:
```
Comment (Calinou) [inline doc/classes/Input.xml:129]:
```suggestion
				On Windows, the dictionary contains the following fields:
```

**Review by Calinou — APPROVED** [review]
Tested locally, it works as expected on Windows 11 23H2.

**Testing project:** [test-joypad-vendor.zip](https://github.com/user-attachments/files/17693388/test-joypad-vendor.zip)

> Format for the results below:
>
> ```
> Name { Info }
> ```

Xbox Elite Series 2 (wired):

```
XInput Gamepad: { "xinput_index": 0, "vendor_id": "1118", "product_id": "767", "xinput_name": "Xbox One For Windows" }
```

DualSense (wired):

```
PS5 Controller: { "vendor_id": "19461", "product_id": "58892" }
```

Switch Pro Controller (wired):

```
Nintendo Switch Pro Controller: { "vendor_id": "32261", "product_id": "2336" }
```

Nyxi Warrior (wired):

```
XInput Gamepad: { "xinput_index": 1, "vendor_id": "1118", "product_id": "2850", "xinput_name": "XBOX 360 For Windows" }
```

**Review by AThousandShips — APPROVED** [review]

File: platform/windows/joypad_windows.cpp [hunk 4]
```diff
@@ -302,52 +358,34 @@ void JoypadWindows::close_joypad(int id) {
...
+	for (int i = 0; i < d_joypad_count; i++) {
 		d_joypads[i].confirmed = false;
 	}
 
+	x_joypad_probe_count = 0;
 	dinput->EnumDevices(DI8DEVCLASS_GAMECTRL, enumCallback, this, DIEDFL_ATTACHEDONLY);
 
-	for (int i = 0; i < joypad_count; i++) {
+	for (int i = x_joypad_probe_count; i < XUSER_MAX_COUNT; i++) {
+		this->probe_xinput_joypad();
```
Comment (AThousandShips) [inline platform/windows/joypad_windows.cpp:375]:
```suggestion
		probe_xinput_joypad();
```

File: doc/classes/Input.xml [hunk 5]
```diff
@@ -127,7 +127,10 @@
...
-				[code]xinput_index[/code]: The index of the controller in the XInput system.
+				[code]xinput_index[/code]: The index of the controller in the XInput system. Is undefined for DirectInput devices.
```
[the remark this answers is not in the document]
Comment (Mickeon) [reply doc/classes/Input.xml:130]: I'd like to keep it brief with just "Undefined" personally.

**Comment by MJacred** [comment]
Applied all requested changes. Thanks for the feedback, everybody!

What do you all think about the change I mentioned in the issue's description regarding changing the joy name for XInput devices? I.e. not adding `xinput_name` to the dictionary. The "name" `XInput Gamepad` is completely redundant. We already have the XInput "GUID" `__XINPUT_DEVICE__`

As an example, for "Xbox Elite Series 2 (wired)", you would get
```
Xbox One For Windows: { "xinput_index": 0, "vendor_id": "1118", "product_id": "767" }
```

instead of 
```
XInput Gamepad: { "xinput_index": 0, "vendor_id": "1118", "product_id": "767", "xinput_name": "Xbox One For Windows" }
```

**Comment by MJacred** [comment]
@bruvzg: Could you take a look at this PR, please? And also give feedback regarding the proposal to not override the joy names for XInput devices to "XInput Gamepad", but use the real name which we can now fetch (see [here](https://github.com/godotengine/godot/pull/98861#issuecomment-2470513052)).

**Comment by bruvzg** [comment]
> And also give feedback regarding the proposal to not override the joy names for XInput devices to "XInput Gamepad", but use the real name which we can now fetch (see https://github.com/godotengine/godot/pull/98861#issuecomment-2470513052).

Fully agree with it, `XInput Gamepad` as a name is redundant, and using actual name is better. I'll test PR tomorrow, code looks fine.

**Comment by MJacred** [comment]
Regarding testing the issue: I replaced the test project in the issue description with one that reflects the latest change (i.e. XInput joypad name).

I also noticed just now that the name set in the mapping db replaces the name coming from the OS/driver in `joy_connection_changed(…, p_connected = true, …)`. Which makes sense as it improves naming consistency across platforms - though it won't work for xinput devices and (possibly) devices for regular xr on android (which makes use of fallback mapping).  
Anyway… `Input::add_joy_mapping(…, p_update_existing = true)` _didn't_ update the name to reflect the new mapping. Which is a bug.  
_Long story short: I added `_set_joypad_mapping` to handle mapping changes for a specific `Joypad` in one place._

**Review by bruvzg — APPROVED** [review]

**Comment by Repiteo** [comment]
Thanks!
